src/server/injected/injectedScript.ts

```diff
@@ -594,36 +599,79 @@ export class InjectedScript {
-    }
-    if (hitElement === element)
-      return 'done';
-    const hitTargetDescription = this.previewNode(hitParents[0]);
-    // Root is the topmost element in the hitTarget's chain that is not in the
-    // element's chain. For example, it might be a dialog element that overlays
-    // the target.
-    let rootHitTargetDescription: string | undefined;
-    while (element) {
-      const index = hitParents.indexOf(element);
-      if (index !== -1) {
-        if (index > 1)
-          rootHitTargetDescription = this.previewNode(hitParents[index - 1]);
-        break;
+
+    const events: string[] = [];
+    if (action === 'hover')
+      events.push('mousemove');
+    else if (action === 'tap')
+      events.push('pointerdown', 'pointerup', 'touchstart', 'touchend', 'touchcancel');
+    else
+      events.push('mousedown', 'mouseup', 'pointerdown', 'pointerup', 'click', 'auxclick', 'dblclick', 'contextmenu');
+
+    let result: 'done' | { hitTargetDescription: string } | undefined;
+
+    const listener = (event: Event) => {
+      // Determine whether we hit the target element on the first event.
+      if (result === undefined) {
+        const hitNode = event.composedPath()[0] as Node;
+        let hitElement = hitNode.nodeType === Node.ELEMENT_NODE ? (hitNode as Element) : (hitNode.parentElement || undefined);
+        const hitParents: Element[] = [];
+        while (hitElement && hitElement !== element) {
+          hitParents.push(hitElement);
+          hitElement = parentElementOrShadowHost(hitElement);
+        }
+        if (hitElement === element) {
+          result = 'done';
+        } else {
+          const hitTargetDescription = this.previewNode(hitParents[0]);
+          // Root is the topmost element in the hitTarget's chain that is not in the
+          // element's chain. For example, it might be a dialog element that overlays
+          // the target.
+          let rootHitTargetDescription: string | undefined;
+          while (element) {
+            const index = hitParents.indexOf(element);
+            if (index !== -1) {
+              if (index > 1)
+                rootHitTargetDescription = this.previewNode(hitParents[index - 1]);
+              break;
+            }
+            element = parentElementOrShadowHost(element);
+          }
+          if (rootHitTargetDescription)
+            result = { hitTargetDescription: `${hitTargetDescription} from ${rootHitTargetDescription} subtree` };
+          else
+            result = { hitTargetDescription };
+        }
       }
-      element = parentElementOrShadowHost(element);
+      if (blockAllEvents || result !== 'done') {
+        event.preventDefault();
+        event.stopPropagation();
+      }
+    };
+
+    const stop = () => {
+      this._removeHitTargetListeners();
+      return result!;
+    };
+
+    this._removeHitTargetListeners();
+    for (const eventName of events) {
+      window.addEventListener(eventName, listener, { capture: true });
```

Comment:
Event listeners are fired in the order they were added so I believe these should be added in evaluateOnNewDocument and then just enabled as needed here. Otherwise an existing window listener will get the bad events. Please cover with a test.